Create correlation matrices for RSEM TPM polya and stranded datasets

e-t-k opened the pull request:
<!--Hi there, thanks for your contribution! Please take a moment to fill out this template to facilitate the review of your pull request.-->
I'm pulling this as a draft request because I'm adding two files,  `01-correlation-matrix.py` and a`utils.py` module with shared functions. Does it make sense to merge the utils as standalone, even though it's a helper that doesn't run anything by itself? Or is it reasonable to review both these files together? Thank you!

### Purpose/implementation Section
#### What scientific question is your analysis addressing?

What is the correlation between samples using RSEM TPM gene expression? We are interested in correlation within the polyA and ribodeplete (stranded) data sets; we don't calculate correlation between the datasets as they are not comparable.

#### What was your approach?

 -  Normalize the RSEM TPM gene expression values by log<sub>2</sub>(TPM + 1). (This is not strictly necessary for the correlation step but will be used in later steps of the analysis.)

Then, apply the tumormap-creation process used in [Vaske et al. (2019) "Comparative Tumor RNA Sequencing Analysis..."](https://jamanetwork.com/journals/jamanetworkopen/fullarticle/2753519), which comprises the following steps:

- Apply filters to the expression matrix to reduce noise that obscures cohort-wide patterns.
	- Drop all genes that are expressed in fewer than 20% of samples
	- Sort the genes by variance and filter out the 20% least varying genes.

- Calculate the [Spearman correlations](https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient) between each pair of samples. 
	


#### What GitHub issue does your pull request address?
#229 (This is the first of 4+ pulls planned for this issue) 



### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?
- Any pull request instructions I've missed or appropriate organization of code/output files.
- RDS file input/output (in utils.py) and the log2 conversion step

I was able to test the filtering and spearman correlations steps in prepare_expression() and calculate_correlation() on Treehouse data sets and results were identical to the output generated by our internal pipeline.


#### Is there anything that you want to discuss further?
- Is RDS an appropriate format for output/scratch files? do you have a standardized format?



#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?
Not yet - no figures/tables generated



### Results

#### What types of results are included (e.g., table, figure)?
No result files yet; not sure if all-by-all file is worthwhile to include as a supplementary table.
 Scratch files include:
 
All-by-all Spearman correlation matrix for each of the polya and stranded RSEM TPM datasets:
```
rsem-tpm-polya-all_by_all_correlations.rds,
rsem-tpm-stranded-all_by_all_correlations.rds
```

List of genes surviving the filters for each dataset:
```
rsem-tpm-polya-filtered_genes_to_keep.rds
rsem-tpm-stranded-filtered_genes_to_keep.rds
```


#### What is your summary of the results?

No analysis performed on these matrices yet - they will also be used as inputs for next step of this analysis


#### Reproducibility Checklist

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

e-t-k commented:
Shall I go ahead and merge your current master into my branch, or is that something one of you prefers to do? (The [contributing page ](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/master/CONTRIBUTING.md#filing-a-pull-request-from-your-own-branch) suggests I keep my fork synced with master only after the first pull is accepted, which is why I haven't done it already.)

In any case, the current CI fail appears to be because my [download-ci-files.sh](https://github.com/UCSC-Treehouse/OpenPBTA-analysis/blob/create-correlation-matrices/scripts/download-ci-files.sh#L43) script is out of date.

jaclyn-taroni commented:
Hi @e-t-k, thanks for filing this! Please feel free to pull master into this branch. I also took a quick look at the files changed here and it seems appropriate to add both files in a single pull request, so you are all set to mark as ready for review when you’re ready.

e-t-k commented:
@jaclyn-taroni thanks for your guidance! It's ready to review now (and thanks for restarting the CI, if that was you who gave it a kick)

jaclyn-taroni commented:
Happy to help! I've asked @jashapiro to review.

jashapiro commented on analyses/comparative-RNASeq-analysis/01-correlation-matrix.py:
```diff
@@ -0,0 +1,107 @@
+
+    # Run pearson metric on ranked genes. For pairwise_distances we need samples=rows so transpose.
+    print_v("Running pairwise distances")
+    x_corr = 1 - sklp.pairwise_distances(X=filtered_rank_transformed_df.transpose(), metric="correlation")
```
I'm curious here if there is a reason not to use `scipy.stats.spearmanr()`? I would think it should have the same result, so it shouldn't really matter, but I'm curious if there might be some issue that this method addresses that led you to use it here?

e-t-k replied:
Yeah, I also expect the result should be the same. This part of the code is copied pretty directly from our [UCSC TumorMap library]( https://github.com/Stuartlab-UCSC/hexmap-data/blob/master/calc/compute_sparse_matrix.py#L332) so that we can potentially generate TumorMaps from these data.

jashapiro commented on analyses/comparative-RNASeq-analysis/01-correlation-matrix.py:
```diff
@@ -0,0 +1,107 @@
+
+def prepare_expression(input_matrix,
+    proportion_unexpressed=0.8,
+    variance_filter_level=0.2,
```
Should these be added as optional arguments that can be passed from the command line?

e-t-k replied:
Done.

e-t-k commented on analyses/comparative-RNASeq-analysis/utils.py:
```diff
@@ -0,0 +1,43 @@
+    construct path to file in that location"""
+    valid_locations={"data":"../../data",
+        "scratch":"../../scratch",
+        "results": "results"
```
Thanks for this insight.  I do want to have the data/scratch/result dirs handled in some centralized way;  what I'm thinking of now is to have them be arguments to my wrapper script (defaulting to current arrangement) - does that seem like a reasonable improvement to you?

```
optional arguments:
  -h, --help            show this help message and exit
  --verbose
  --results RESULTS     Path to dir storing final result files.
  --scratch SCRATCH     Path to scratch dir.
```

jashapiro replied:
That seems a great solution, and probably what I would lean toward doing! It makes it easy to change for all the wrapped scripts, and nobody should have to look at the underlying code to see what needs changing.

e-t-k commented on analyses/comparative-RNASeq-analysis/01-correlation-matrix.py:
```diff
@@ -0,0 +1,107 @@
+    datasets_to_process = [
+        ("pbta-gene-expression-rsem-tpm.polya.rds", "rsem-tpm-polya-"),
+        ("pbta-gene-expression-rsem-tpm.stranded.rds", "rsem-tpm-stranded-")
+    ]
```
Done. I plan to add the shell script with the next step of the analysis.

e-t-k commented on analyses/comparative-RNASeq-analysis/utils.py:
```diff
@@ -0,0 +1,43 @@
+    filepath = get_filepath(filename, location)
+    pyreadr.write_rds(filepath, df)
+    # Pyreadr.write_rds fails silently when permissions prevent file write,
+    # so trigger an error if our file isn't actually there
```
Agreed...

e-t-k commented:
Thank you @jashapiro ! I believe I've addressed your comments in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/323/commits/b208c8bd5933ad7945a56ed1c93ae43f65bf0091.

jashapiro commented on analyses/comparative-RNASeq-analysis/01-correlation-matrix.py:
```diff
@@ -0,0 +1,113 @@
+    p = argparse.ArgumentParser()
+    p.add_argument("--verbose", action="store_true")
+    p.add_argument("input_path", metavar="input-path", help="Path to input Rds file.")
+    p.add_argument("--scratch", help="Path to scratch dir.", default=os.path.join("..", "..", "scratch"))
```
```suggestion
    p.add_argument("--scratch", 
                   help="Path to scratch dir.", 
                   default=os.path.join("..", "..", "scratch"))
```
